serverkeyapi/internal/api.go

```diff
@@ -45,51 +79,153 @@ func (s *ServerKeyAPI) FetchKeys(
 	// because the caller gives up waiting.
 	ctx := context.Background()
 	results := map[gomatrixserverlib.PublicKeyLookupRequest]gomatrixserverlib.PublicKeyLookupResult{}
+	origRequests := map[gomatrixserverlib.PublicKeyLookupRequest]gomatrixserverlib.Timestamp{}
+	for k, v := range requests {
+		origRequests[k] = v
+	}
 	now := gomatrixserverlib.AsTimestamp(time.Now())
-	// First consult our local database and see if we have the requested
+
+	// First, check if any of these key checks are for our own keys. If
+	// they are then we will satisfy them directly.
+	for req := range requests {
+		if req.ServerName == s.Cfg.Matrix.ServerName {
+			// We found a key request that is supposed to be for our own
+			// keys. Remove it from the request list so we don't hit the
+			// database or the fetchers for it.
+			delete(requests, req)
+
+			// Look up our own keys.
+			request := &api.QueryLocalKeysRequest{}
+			response := &api.QueryLocalKeysResponse{}
+			if err := s.QueryLocalKeys(ctx, request, response); err != nil {
+				return nil, err
+			}
+
+			// Depending on whether the key is expired or not, we'll need
+			// to write slightly different
+			if verifyKeys, ok := response.ServerKeys.VerifyKeys[req.KeyID]; ok {
+				// The key is current.
+				results[req] = gomatrixserverlib.PublicKeyLookupResult{
+					VerifyKey:    verifyKeys,
+					ExpiredTS:    gomatrixserverlib.PublicKeyNotExpired,
+					ValidUntilTS: response.ServerKeys.ValidUntilTS,
+				}
+			} else if verifyKeys, ok := response.ServerKeys.OldVerifyKeys[req.KeyID]; ok {
+				// The key is expired.
+				results[req] = gomatrixserverlib.PublicKeyLookupResult{
+					VerifyKey:    verifyKeys.VerifyKey,
+					ExpiredTS:    verifyKeys.ExpiredTS,
+					ValidUntilTS: gomatrixserverlib.PublicKeyNotValid,
+				}
+			}
+		}
+	}
+
+	// Then consult our local database and see if we have the requested
 	// keys. These might come from a cache, depending on the database
 	// implementation used.
 	if dbResults, err := s.OurKeyRing.KeyDatabase.FetchKeys(ctx, requests); err == nil {
-		// We successfully got some keys. Add them to the results and
-		// remove them from the request list.
+		// We successfully got some keys. Add them to the results.
 		for req, res := range dbResults {
-			if !res.WasValidAt(now, true) {
-				// We appear to be past the key validity. Don't return this
-				// key with the results.
-				continue
-			}
 			results[req] = res
```

Comment:
We're now adding invalid keys to `results`. Why?

Reply:
The logic in this PR aims to make a best-effort attempt to find the latest keys, but if it can't find newer keys, it will return the ones that we know about. This the only way to satisfy verifications for when the event happened in the past and the key was never renewed (i.e. the origin server is now dead).

Reply:
Please add a comment to this effect somewhere, I can't see it anywhere. It's mainly relevant for `handleDatabaseKeys`